org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/refactoring2/participant/ResourceRelocationProcessor.java

```diff
@@ -0,0 +1,186 @@
+		if (uriChanges.isEmpty()) {
+			return null;
+		}
+
+		SubMonitor subMonitor = SubMonitor.convert(pm);
+		// Declaring the task and its effort in 'SubMonitor.convert(...)' doesn't yield the expected UI updates
+		// so let's do it separately; the total effort of '5' is chosen for weighting the subsequent efforts
+		subMonitor.beginTask("Preparing the refactoring...", 5);
+
+		IChangeSerializer changeSerializer = changeSerializerProvider.get();
+		ResourceSet resourceSet = resourceSetProvider.get(project);
+
+		ResourceRelocationContext context = new ResourceRelocationContext(type, uriChanges, issues, changeSerializer, resourceSet);
+		boolean persistedIndexUsageRequested = isPersistedIndexUsageRequested(context);
+
+		// TODO check preconditions like all editors being saved if 'persistedIndexUsageRequested' == true
+
+		initializeResourceSet(persistedIndexUsageRequested, context);
+		executeParticipants(context, subMonitor.split(1));
+
+		ChangeConverter changeConverter = changeConverterFactory.create(name, //
+				(it) -> {
+					return (!(it instanceof MoveResourceChange || it instanceof RenameResourceChange)
+							|| !excludedResources.contains(it.getModifiedElement()));
+				}, issues);
+
+		SubMonitor modificationApplicationMonitor = subMonitor.split(4); // remaining effort is assigned to 'changeSerializer's work
+		changeSerializer.setProgressMonitor(modificationApplicationMonitor);
+		changeSerializer.applyModifications(changeConverter);
+		modificationApplicationMonitor.done();
+		return changeConverter.getChange();
+	}
+
+	/**
+	 * @since 2.18
+	 */
+	protected boolean isPersistedIndexUsageRequested(ResourceRelocationContext context) {
+		List<? extends IResourceRelocationStrategy> strategies = strategyRegistry.getStrategies();
+		return strategies.stream().anyMatch(strategy -> strategy.requiresUsageOfPersistedIndex(context));
+	}
+
+	/**
+	 * @since 2.18
+	 */
+	protected void initializeResourceSet(boolean persistedIndexUsageRequested, ResourceRelocationContext context) {
+		if (persistedIndexUsageRequested) {
+			context.getResourceSet().getLoadOptions().put(ResourceDescriptionsProvider.PERSISTED_DESCRIPTIONS, Boolean.TRUE);
+		} else {
+			liveScopeResourceSetInitializer.initialize(context.getResourceSet());
+		}
+	}
+
+	protected void executeParticipants(ResourceRelocationContext context, SubMonitor monitor) {
+		List<? extends IResourceRelocationStrategy> strategies = strategyRegistry.getStrategies();
+		if (context.getChangeType() == ResourceRelocationContext.ChangeType.COPY) {
+			context.getChangeSerializer().setUpdateRelatedFiles(false);
+		}
+
+		monitor.setWorkRemaining(strategies.size());
+
+		for (IResourceRelocationStrategy strategy : strategies) {
+			try {
+				monitor.split(1);
+				strategy.applyChange(context);
+			} catch (OperationCanceledException t) {
+				issues.add(ERROR, "Operation was cancelled while applying resource changes", t);
+				LOG.error(t.getMessage(), t);
+				break;
```

Comment:
@tivervac @szarnekow I have added the `break` here. The original code didn't have it. Please confirm if this is ok.

Reply:
Seems fine to me